lib/wit/dependency.py

```diff
@@ -0,0 +1,158 @@
+        This will also add itself as a dependent of the package
+        """
+        if self.name in packages:
+            self.package = packages[self.name]
+        else:
+            self.package = Package(self.name, repo_paths)
+            self.package.set_source(self.source)
+            packages[self.name] = self.package
+        self.package.add_dependent(self)
+
+    def add_dependent(self, dependent):
+        if dependent not in self.dependents:
+            self.dependents.append(dependent)
+
+    def get_commit_time(self):
+        return datetime.utcfromtimestamp(int(self.package.repo.commit_to_time(
+            self.specified_revision)))
+
+    def manifest(self):
+        return {
+            'name': self.name,
+            'source': self.source,
+            'commit': self.specified_revision,
+        }
+
+    # used before saving to manifests/lockfiles
+    def resolved(self):
+        if self.package is None or self.package.repo is None:
+            raise Exception("Cannot resolve dependency that us unbound to disk")
+        return Dependency(self.name, self.source, self.resolved_rev())
+
+    def resolved_rev(self):
+        if self.package.repo is None or self.package.repo is None:
+            raise Exception("Cannot resolve dependency that is unbound to disk")
+        return self.package.repo.get_commit(self.specified_revision)
+
+    def __repr__(self):
+        return "Dep({})".format(self.tag())
+
+    def short_revision(self):
+        if self.package and self.package.repo:
+            if self.package.repo.is_hash(self.specified_revision):
+                return self.package.repo.get_shortened_rev(self.specified_revision)
+            return self.specified_revision
+        return self.specified_revision[:8]
+
+    def tag(self):
+        return "{}::{}".format(self.name, self.short_revision())
+
+    def get_id(self):
+        return "dep_"+re.sub(r"([^\w\d])", "_", self.tag())
+
+    def crawl_dep_tree(self, wsroot, repo_paths, packages):
+        fancy_tag = "{}::{}".format(self.name, self.short_revision())
+        self.load_package(packages, repo_paths)
+        self.package.load_repo(wsroot)
+        if self.package.repo is None:
+            return {'': "{} \033[91m(missing)\033[m".format(fancy_tag)}
+        if self.package.revision != self.resolved_rev():
+            fancy_tag += "->{}".format(self.package.short_revision())
+            return {'': fancy_tag}
+
+        tree = {'': fancy_tag}
+        subdeps = self.package.get_dependencies()
+        for subdep in subdeps:
+            tree[subdep.get_id()] = subdep.crawl_dep_tree(wsroot, repo_paths, packages)
+        return tree
+
+
+def parse_dependency_tag(s):
+    # TODO Could speed up validation
+    #   - use git ls-remote to validate remote exists
+    #   - use git ls-remote to validate revision for tags and branches
+    #   - if github repo, check if page exists (or if you get 404)
```

Comment:
These comments are obsolete (at least with regards to this code). They might apply elsewhere, otherwise they should be removed

Reply:
I think that we should still eventually use `ls-remote` for resolving non-hash references. Should I move the comment to the top of `gitrepo.py`?

Reply:
Yeah that would make more sense

Reply:
Fixed by https://github.com/sifive/wit/pull/96/commits/ad53f1d1676df53397c121853a28620118f67f2c